fhir-persistence-schema/docs/DB2MultiTenancy.md

```diff
@@ -0,0 +1,140 @@
+
+The tables have supporting indicies and privileges. 
+
+**Table: TENANTS**
+
+The Tenants table maps a given tenant to the Multi-Tenant ID (MT_ID) for the given tenant.  The MT_ID is subsequently used to restrict row access based on the matching value. 
+
+| Column Name | Data Type | Length | Null | Purpose |
+|----------|---------------|---------------|---------------|---------------------------------------------|
+| MT_ID | Integer | 4 | No | Unique ID identifying the Tenant's ID|
+| TENANT_NAME | VARCHAR | 36 | No | The name of the tenant |
+| TENANT_STATUS | VARCHAR | 16 | No | The status of the Tenant|
+
+MT_ID is the value used to assign ROW Access Control in the Multi-Tenant schema. The MT_ID is assigned from the FHIR_ADMIN.TENANT_SEQUENCE.  The value is to 1 to 9999, with no cycles. 
+
+Tenant Status is one of the following: 
+| Status | Purpose |
+|----------|---------------------------------------------|
+| PROVISIONING | Indicates the tenant schema is in the process of adding a set of tenant tables/partitions |
+| ALLOCATED | Indicates the process of creating the tenant schema is finished, and the schema is ready for use. Only allocated tenants are available for use in the `fhir-persistence-jdbc` code |
+| FROZEN | Indicates the previous tenant data is not to be used, and the tenant data is in the process of being removed |
+| DROPPED | Indicates the previous tenant data is now removed |
+| FREE | Indicates a set of tables is provisioned, and available to be allocated to a specific tenant |
+
+**Table: TENANT_KEYS**
+
+The Tenant Keys table stores a hashed version of the tenant specific key. Upon creating a session and calling the SET_TENANT procedure, the SET_TENANT queries the incoming key against the TENANT_KEYS database to map to a specific tenant MT_ID. 
+
+| Column Name | Data Type | Length | Null | Purpose |
+|----------|---------------|---------------|---------------|---------------------------------------------|
+| TENANT_KEY_ID | Integer | 4 | No | The ID is from the sequence, and is used to identify the unique entry in the row |
+| MT_ID | Integer | 4 | No | Matches to the MT_ID in the TENANTS table |
+| TENANT_SALT | VARCHAR | 44 | No | A 32 byte random value used as a salt when computing the SHA-256 tenant hash |
+| TENANT_HASH | VARBINARY | 32 | No | Each tenant's data is encrypted. |
+
+The tenant_key is a 32 byte random value created when the tenant is provisioned. This is a secret, and must be passed by the client to the database in order to obtain access to a tenant. A tenant may have multiple tenant keys to support key rotation. An example of the tenant key is `GQUHRi0ETFw0Xe+p1VtVDki4JFZs885bGM35hXRyq/E=`
+
+Multiple keys per tenant are supported, and key rotation includes adding, and then subsequently removing prior keys. 
+
+**Table: VERSION_HISTORY**
+
+The VERSION_HISTORY table enables Tenant version isolation, so upgrades and patches are separately applied to each tenant's schema. 
+
```

Comment:
on each tenant's schema or on "each tenant schema"?
e.g, we have FHIRDATA and FHIRDATA2 schemas in CDT, so we have 2 tenant schemas there. and tenants can be created in each of these 2 tenant schema...  so, we don't really update the schema per tenant ...

Reply:
The code actually supports this feature.  We just chose not to take this step in CDT.

Reply:
en... this is interesting... Is it "TableSpace" as OBJECT_TYPE and tenant name or id as OBJECT_NAME if we choose to upgrade schema for a certain tenant only?   
BTW, I noticed that schemaName is not really used in allocateTenant(String adminSchemaName, String schemaName, String tenantName, String tenantKey, String tenantSalt, String idSequenceName) ...

Reply:
OBJECT_NAME is not a tenant_name.  It's inferred based 

Yeah - that's right allocateTenant actually figures it out in the query.

Reply:
I had the same question and I'm still confused.  The code supports updating the tables on a per-*schema* basis right?  Which might be a single tenant or might be many tenants?

Reply:
>  e.g, we have FHIRDATA and FHIRDATA2 schemas in CDT, so we have 2 tenant schemas there. and tenants can be created in each of these 2 tenant schema... so, we don't really update the schema per tenant ...

Well.  Let me clarify... the code technically supports it.  We're just breaking the model by removing columns and changing indices.  It does support independent updates.  We also support independent schemas.  

I've changed the words and given a few cases where this is true.

Reply:
> en... this is interesting... Is it "TableSpace" as OBJECT_TYPE and tenant name or id as OBJECT_NAME if we choose to upgrade schema for a certain tenant only?
an OBJECT_TYPE is one of the types

[db2inst1@c1ed9e48d0ab ~]$ db2 "SELECT DISTINCT OBJECT_TYPE FROM FHIR_ADMIN.VERSION_HISTORY"

OBJECT_TYPE
----------------
PROCEDURE
SEQUENCE
TABLE
TABLESPACE
VARIABLE

  5 record(s) selected.

We only pass in the tenant_id and run an updateSchema action. 

> BTW, I noticed that schemaName is not really used in allocateTenant(String adminSchemaName, String schemaName, String tenantName, String tenantKey, String tenantSalt, String idSequenceName) ...
Yes, this is correct... the actions taken in each part of the fhir-persistence-schema does not necessarily use all inputs.

Reply:
> I had the same question and I'm still confused. The code supports updating the tables on a per-_schema_ basis right? Which might be a single tenant or might be many tenants?

It's actually not the design... it's just how we've done it that has broken the contract.  Robin did an always add approach, don't take away... This has resulted in an update independent approach.  We can support it, however, we haven't followed it. 

I've elaborated in the document explaining.